expression/builtin_arithmetic.go

```diff
@@ -646,51 +648,149 @@ func (s *builtinArithmeticIntDivideDecimalSig) evalInt(row []types.Datum) (int64
+		return sig.setSelf(sig), nil
+	} else {
+		bf := newBaseBuiltinFuncWithTp(args, ctx, tpInt, tpInt, tpInt)
+		if mysql.HasUnsignedFlag(tpA.Flag) {
+			bf.tp.Flag |= mysql.UnsignedFlag
+		}
+		sig := &builtinArithmeticModIntSig{baseIntBuiltinFunc{bf}}
+		return sig.setSelf(sig), nil
+	}
 }
 
-type builtinArithmeticSig struct {
-	baseBuiltinFunc
-	op opcode.Op
+type builtinArithmeticModRealSig struct {
+	baseRealBuiltinFunc
 }
 
-func (s *builtinArithmeticSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := s.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
+func (s *builtinArithmeticModRealSig) evalReal(row []types.Datum) (float64, bool, error) {
+	sc := s.ctx.GetSessionVars().StmtCtx
+	b, isNull, err := s.args[1].EvalReal(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+
+	if b == 0 {
+		return 0, true, types.ErrDivByZero
 	}
+
+	a, isNull, err := s.args[0].EvalReal(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+
+	return math.Mod(a, b), false, nil
+}
+
+type builtinArithmeticModDecimalSig struct {
+	baseDecimalBuiltinFunc
+}
+
+func (s *builtinArithmeticModDecimalSig) evalDecimal(row []types.Datum) (*types.MyDecimal, bool, error) {
 	sc := s.ctx.GetSessionVars().StmtCtx
-	a, err := types.CoerceArithmetic(sc, args[0])
-	if err != nil {
-		return d, errors.Trace(err)
+	a, isNull, err := s.args[0].EvalDecimal(row, sc)
+	if isNull || err != nil {
+		return nil, isNull, errors.Trace(err)
+	}
+	b, isNull, err := s.args[1].EvalDecimal(row, sc)
+	if isNull || err != nil {
+		return nil, isNull, errors.Trace(err)
 	}
+	c := &types.MyDecimal{}
+	err = types.DecimalMod(a, b, c)
+	return c, err == types.ErrDivByZero, errors.Trace(err)
+}
 
-	b, err := types.CoerceArithmetic(sc, args[1])
-	if err != nil {
-		return d, errors.Trace(err)
+type builtinArithmeticModIntSig struct {
+	baseIntBuiltinFunc
+}
+
+func (s *builtinArithmeticModIntSig) evalInt(row []types.Datum) (val int64, isNull bool, err error) {
+	sc := s.ctx.GetSessionVars().StmtCtx
+
+	b, isNull, err := s.args[1].EvalInt(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
 	}
-	a, b, err = types.CoerceDatum(sc, a, b)
-	if err != nil {
-		return d, errors.Trace(err)
+
+	if b == 0 {
+		return 0, true, types.ErrDivByZero
 	}
-	if a.IsNull() || b.IsNull() {
-		return
+
+	a, isNull, err := s.args[0].EvalInt(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
 	}
 
-	switch s.op {
-	case opcode.Mod:
-		return types.ComputeMod(sc, a, b)
-	default:
-		return d, errInvalidOperation.Gen("invalid op %v in arithmetic operation", s.op)
+	var ret int64
+	isLHSUnsigned := mysql.HasUnsignedFlag(s.args[0].GetType().Flag)
+	isRHSUnsigned := mysql.HasUnsignedFlag(s.args[1].GetType().Flag)
+
+	switch {
+	case isLHSUnsigned && isRHSUnsigned:
+		ret = int64(uint64(a) % uint64(b))
+	case isLHSUnsigned && !isRHSUnsigned:
+		if b < 0 {
+			ret = int64(uint64(a) % uint64(-b))
```

Comment:
if b == MIN_INT64, -b may overflow?

Reply:
It won't overflow. I've try it.

Reply:
what will `select cast(1 as unsigned) mod -9223372036854775808` after this commit?
In MySQL:
``` sql
mysql> select cast(1 as unsigned) mod -9223372036854775808;
Field   1:  `cast(1 as unsigned) mod -9223372036854775808`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       LONGLONG
Collation:  binary (63)
Length:     20
Max_length: 1
Decimals:   0
Flags:      UNSIGNED BINARY NUM


+----------------------------------------------+
| cast(1 as unsigned) mod -9223372036854775808 |
+----------------------------------------------+
|                                            1 |
+----------------------------------------------+
1 row in set (0.04 sec)
```